libs/wingc/src/inflight_transform.rs

```diff
@@ -0,0 +1,277 @@
+		self.curr_phase = prev_phase;
+		new_node
+	}
+
+	fn fold_class(&mut self, node: Class) -> Class {
+		let prev_is_inside_resource = self.is_inside_resource;
+		self.is_inside_resource = true;
+		let new_node = fold::fold_class(self, node);
+		self.is_inside_resource = prev_is_inside_resource;
+		new_node
+	}
+
+	fn fold_expr(&mut self, expr: Expr) -> Expr {
+		// Inflight closures that are themselves defined inflight do not need
+		// to be transformed, since they are not created in preflight.
+		if self.curr_phase == Phase::Inflight {
+			return expr;
+		}
+
+		// If we encounter a non-inflight closure, we can don't need to
+		// transform it into a resource, but we still want to recurse
+		// in case its body contains any inflight closures.
+		if let ExprKind::FunctionClosure(ref func_def) = expr.kind {
+			if func_def.signature.phase != Phase::Inflight {
+				return fold::fold_expr(self, expr);
+			}
+		}
+
+		match expr.kind {
+			ExprKind::FunctionClosure(func_def) => {
+				self.inflight_counter += 1;
+
+				let resource_name = Symbol {
+					name: format!("$Inflight{}", self.inflight_counter),
+					span: expr.span.clone(),
+				};
+				let handle_name = Symbol {
+					name: "handle".to_string(),
+					span: expr.span.clone(),
+				};
+				let parent_this_name = Symbol {
+					name: PARENT_THIS_NAME.to_string(),
+					span: expr.span.clone(),
+				};
+				let this_name = Symbol {
+					name: "this".to_string(),
+					span: expr.span.clone(),
+				};
+
+				let resource_type_annotation = TypeAnnotation {
+					kind: TypeAnnotationKind::UserDefined(UserDefinedType {
+						root: resource_name.clone(),
+						fields: vec![],
+					}),
+					span: expr.span.clone(),
+				};
+
+				let resource_fields: Vec<ClassField> = vec![];
+				let resource_init_params: Vec<FunctionParameter> = vec![];
+
+				let new_func_def = if self.is_inside_resource {
+					// transform all occurrences of "this" into "__parent_this"
+					let mut this_transform = ThisTransformer;
+					this_transform.fold_function_definition(func_def)
+				} else {
+					func_def
+				};
+
+				// resource_def :=
+				// ```
+				// resource {
+				//   init(<resource_init_params>) {<resource_init_body>}
+				//   inflight handle() {
+				//     <transformed_def>
+				//   }
+				// }
+				// ```
+				let resource_def = Stmt {
+					kind: StmtKind::Class(Class {
+						name: resource_name.clone(),
+						is_resource: true,
+						initializer: Initializer {
+							signature: FunctionSignature {
+								parameters: resource_init_params,
+								return_type: Some(Box::new(resource_type_annotation.clone())),
+								phase: Phase::Preflight,
+							},
+							statements: Scope::new(vec![], expr.span.clone()),
+							span: expr.span.clone(),
+						},
+						fields: resource_fields,
+						implements: vec![],
+						parent: None,
+						methods: vec![(handle_name.clone(), new_func_def)],
+						inflight_initializer: None,
+					}),
+					idx: 1,
+					span: expr.span.clone(),
```

Comment:
There's an interesting implementation question here of, "ok, if we're introducing random new AST nodes to the tree (that could potentially get type checked and so on), what `span` do we assign them?" Should it just be a "global" span?

Here I'm trying the approach of "let's just inherit the span of the nearest parent AST node that we're adding ourselves to", which happens to be the span of the original inflight closure. I'm not sure if this is correct, but it seemed like a decent option for making sure any unexpected errors are localized near their source.

Reply:
I think inheriting the original span makes sense to me